Source/WebCore/dom/FullscreenManager.cpp

```diff
@@ -248,40 +216,33 @@ void FullscreenManager::requestFullscreenForElement(Ref<Element>&& element, RefP
 
         // 5. Return, and run the remaining steps asynchronously.
         // 6. Optionally, perform some animation.
         m_areKeysEnabledInFullscreen = hasKeyboardAccess;
-        document->eventLoop().queueTask(TaskSource::MediaElement, [this, weakThis = WTFMove(weakThis), promise = WTFMove(promise), element = WTFMove(element), completionHandler = WTFMove(completionHandler), handleError = WTFMove(handleError), identifier, mode] () mutable {
-            if (!weakThis) {
-                if (promise)
-                    promise->reject(Exception { ExceptionCode::TypeError });
-                completionHandler(false);
-                return;
-            }
+        document->eventLoop().queueTask(TaskSource::MediaElement, [this, weakThis = WTFMove(weakThis), element = WTFMove(element), completionHandler = WTFMove(completionHandler), handleError = WTFMove(handleError), identifier, mode] () mutable {
+            if (!weakThis)
+                return completionHandler(Exception { ExceptionCode::TypeError });
 
             RefPtr page = this->page();
-            if (!page || (this->document().hidden() && mode != HTMLMediaElementEnums::VideoFullscreenModeInWindow) || m_pendingFullscreenElement != element.ptr() || !element->isConnected()) {
-                handleError("Invalid state when requesting fullscreen."_s, EmitErrorEvent::Yes, WTFMove(element), WTFMove(promise), WTFMove(completionHandler));
-                return;
-            }
+            if (!page || (this->document().hidden() && mode != HTMLMediaElementEnums::VideoFullscreenModeInWindow) || m_pendingFullscreenElement != element.ptr() || !element->isConnected())
+                return handleError("Invalid state when requesting fullscreen."_s, EmitErrorEvent::Yes, WTFMove(completionHandler));
 
             // Reject previous promise, but continue with current operation.
             if (m_pendingPromise) {
                 ERROR_LOG(identifier, "Pending operation cancelled by requestFullscreen() call.");
-                m_pendingPromise->reject(Exception { ExceptionCode::TypeError, "Pending operation cancelled by requestFullscreen() call."_s });
+                m_pendingPromise.reject(Exception { ExceptionCode::TypeError, "Pending operation cancelled by requestFullscreen() call."_s });
             }
-
-            m_pendingPromise = WTFMove(promise);
+            m_pendingPromise = WTFMove(completionHandler);
 
             INFO_LOG(identifier, "task - success");
 
-            page->chrome().client().enterFullScreenForElement(element, mode);
-            completionHandler(true);
+            page->chrome().client().enterFullScreenForElement(element, mode, [weakPromise = WeakPtr { m_pendingPromise }] (ExceptionOr<void> result) {
+                if (weakPromise)
+                    weakPromise->rejectOrResolve(WTFMove(result));
+            });
```

Comment:
I don't really understand the benefit of making `FullscreenPromise` a `CanMakeWeakPtr` when the manager itself is already a `CanMakeWeakPtr` and is passed into this lambda as a `WeakPtr` itself. Especially since the lifetime of a `FullscreenPromise` is tied to the lifetime of the `FullscreenManager` itself.

Reply:
Before this change, there was a possibility that enterFullScreen would start, then exitFullScreen would start and reject the promise, then enterFullScreen would complete and reject the wrong promise.  Keeping a WeakPtr to the promise itself prevents this possible confusion.

Reply:
I didn't actually fix this problem, so I'm going with your suggestion.  It is better.